modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/NodePageStream.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache.query.reducer;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.UUID;
+import java.util.function.BiConsumer;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.internal.processors.cache.query.GridCacheQueryAdapter;
+import org.apache.ignite.internal.util.typedef.F;
+
+/**
+ * Stream over single node.
+ */
+public class NodePageStream<R> extends PageStream<R> {
+    /** */
+    private final UUID nodeId;
+
+    /** */
+    private R head;
+
+    /** */
+    protected NodePageStream(GridCacheQueryAdapter qry, Object queueLock, long timeoutTime,
+        UUID nodeId, BiConsumer<Collection<UUID>, Boolean> reqPages) {
+        super(qry, queueLock, timeoutTime, new HashSet<>(F.asList(nodeId)), reqPages);
+
+        this.nodeId = nodeId;
+    }
+
+    /**
+     * @return Head of stream, that is last item returned with {@link #next()}.
+     */
+    public R head() {
+        return head;
+    }
+
+    /**
+     * @return Head of stream and then clean it.
+     */
+    public R get() {
+        R ret = head;
+
+        head = null;
+
+        return ret;
+    }
```

Comment:
This method looks useless.
Al the logic could be done in next() method.

Reply:
You're correct. I wanted to separate logic of filling head() and cleaning(), but actually there is no need to do it.

Reply:
`
@Override public boolean hasNext() throws IgniteCheckedException {
        if (head != null)
            return true;
       else {
            if (!super.hasNext())
                return false;
            else
                head = super.next();   
       }
}
`

Reply:
Agree, moving fetch value from next() to hasNext() resolves this. Added.